vllm/v1/core/scheduler.py

```diff
@@ -648,6 +719,10 @@ def _free_request(self, request: Request) -> None:
         self._cached_reqs_data.pop(request.request_id, None)
         del self.requests[request.request_id]
         self.finished_req_ids.add(request.request_id)
+        if request.use_guided_decoding:
+            # NOTE: grammar should NOT be None
+            # if use_guided_decoding is True
+            request.grammar.reset()  # type: ignore[union-attr]
```

Comment:
This probably isn't necessary. Each request now gets its own matcher instance. Each matcher shares a common, cached, compiled grammar. We're resetting the matcher here, but it's not reused so it's probably just a waste. Need to run some load tests to be sure, but I think so!

Reply:
```suggestion
```